File: problems/75/solution.cpp
// Accepted: 4ms (31.04%), 9.8MB (33.64%)

class Solution {
    void swap(vector<int>& nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

public:
    // Dijkstra's three-way partitioning algorithm
    void sortColors(vector<int>& nums) {
        int len = nums.size();
        int l = 0;
        int m = 0;
        int u = nums.size();

        // [0, l): 0
        // [l, m): 1
        // [m, u): ?
        // [u, len): 2

        while (m < u) {
            if (nums[m] == 1) {
                ++m;
            } else if (nums[m] == 2) {
                --u;
                swap(nums, m, u);
            } else { // nums[m] == 0
                swap(nums, l, m);
                ++l;
                ++m;
            }
        }
    }
};

File: problems/75/solution.py
# Accepted: 35ms (73.85%), 16.41MB (62.53%)

class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        l = 0
        m = 0
        u = len(nums)

        # [0, l): 0
        # [l, m): 1
        # [m, u): ?
        # [u, len): 2

        while m < u:
            num = nums[m]
            if num == 1:
                m += 1
            elif num == 2:
                u -= 1
                nums[m], nums[u] = nums[u], nums[m]
            else:
                nums[m], nums[l] = nums[l], nums[m]
                l += 1
                m += 1
